ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java

```diff
@@ -188,12 +202,96 @@ public String writeSingle(String key) throws Exception {
 
     @Override
     public List<String> readBulk(List<String> keys) throws Exception {
-        return null;
+        Preconditions.checkArgument(new HashSet<>(keys).size() == keys.size());
+        final KeysAndAttributes keysAndAttributes = generateReadRequests(keys);
+        try {
+            readUntilDone(keysAndAttributes);
+            return keysAndAttributes.getKeys().stream()
+                    .map(Map::toString)
+                    .collect(Collectors.toList());
+        } catch (AmazonServiceException ase) {
+            amazonServiceException(ase);
+            throw ase;
+        } catch (AmazonClientException ace) {
+            amazonClientException(ace);
+            throw ace;
+        }
     }
 
     @Override
-    public List<String> writeBulk(List<String> keys) throws Exception {
-        return null;
+    public List<String> writeBulk(List<String> keys) {
+        Preconditions.checkArgument(new HashSet<>(keys).size() == keys.size());
+        final List<WriteRequest> writeRequests = generateWriteRequests(keys);
+        try {
+            writeUntilDone(writeRequests);
+            return writeRequests.stream()
+                    .map(WriteRequest::getPutRequest)
+                    .map(PutRequest::toString)
+                    .collect(Collectors.toList());
+        } catch (AmazonServiceException ase) {
+            amazonServiceException(ase);
+            throw ase;
+        } catch (AmazonClientException ace) {
+            amazonClientException(ace);
+            throw ace;
+        }
+    }
+
+    private List<WriteRequest> generateWriteRequests(List<String> keys) {
+        return keys.stream()
+                .map(key -> ImmutableMap.of(partitionKeyName, new AttributeValue(key),
+                        ATTRIBUTE_NAME, new AttributeValue(this.dataGenerator.getRandomValue())))
+                .map(item -> new PutRequest().withItem(item))
+                .map(put -> new WriteRequest().withPutRequest(put))
+                .collect(Collectors.toList());
+    }
+
+    private void writeUntilDone(List<WriteRequest> requests) {
+        List<WriteRequest> remainingRequests = requests;
+        BatchWriteItemResult result;
+        do {
+            result = runBatchWriteRequest(remainingRequests);
+            remainingRequests = result.getUnprocessedItems().get(table.getTableName());
+        } while (remainingRequests!= null && remainingRequests.isEmpty());
+    }
+
+    private BatchWriteItemResult runBatchWriteRequest(List<WriteRequest> writeRequests) {
+        //estimate size of requests
+        int totalSize = writeRequests.stream()
```

Comment:
Are you using `totalSize` anywhere?

Reply:
For variable data sizes in config, it is a good practice to self-throttle on the client side based on the expected WCU consumption of each item. I haven't implemented this yet, but the size calculation is a precursor: ceiling(bytes / 1024) = WCU. I can pull this out and save for a later date.

Reply:
fixed